SharpRepository.EfRepository/EfRepositoryBase.cs

```diff
@@ -33,7 +34,7 @@ protected override void AddItem(T entity)
             if (typeof(TKey) == typeof(Guid) || typeof(TKey) == typeof(string))
             {
                 TKey id;
-                if (GetPrimaryKey(entity, out id) && Equals(id, default(TKey)))
+                if (GetPrimaryKey(entity, out id) && GenerateKeyOnAdd && Equals(id, default(TKey)))
```

Comment:
I've pushed a new commit doing it like you suggest in all repositories not depending on ```GetPrimaryKey``` to retrieve the ```TKey id``` value. ```InMemoryRepository``` and a few others must execute ```GetPrimaryKey``` even if ```GenerateKeyOnAdd``` is false.